### PR TITLE
Let users sign in with a known provider before linking a new provider to the user

### PR DESCRIPTION
**Pull request by jorg-vr**
This pull request forces a user to confirm their identity before linking their new sign in method to an existing user.

The steps followed in the default scenario are as followed:
1. An existing users signs in using a new  provider
2. We search for an existing identity for this user. (None can be found because it is the first sign in of this user using this provider)
3. Try to find an existing user in the same institution matching username or email (Which is found)
4. Check if user does not already have an identity for this provider
5. Save identity details and user id in session
6. Ask user to confirm that they have access to this user by signing in using a previously used login method
7. User signs in using known provider
8. Check if identity in session matches current signed in user
9. Add new identity to user
10. Notify user

image
image

I also added the linked sign in methods to the user detail screen
image

### Misc
- I created a separate user_short partial for reuse (It is only used at a single place at the moment)
- I removed a failing test which was made specifically for merging token users into real users. As discussed this usecase does not happen
- flash messages can now contain multiple links instead of one

Closes #3668 .

### CODE REVIEW

**Review by bmesuere — CHANGES_REQUESTED** [review]
I would prefer this to be on a separate page or at least a modal instead of a flash/alert. This is an important action for the user and it feels somewhat wrong to stash this away like this.

If we make linking more transparent like this, it might also be  a good idea to list the linked providers on the profile page of a user?

File: config/locales/devise.nl.yml [hunk 1]
```diff
@@ -32,6 +32,11 @@ nl:
       wrong_provider_extra: "Aanmelden via %{user_provider_type}."
       linked: Omdat je nog niet eerder op Dodona inlogde vanaf Ufora, konden we de gevraagde actie niet uitvoeren. Nu we je UGent-account gekoppeld hebben, zou dit wel moeten lukken. Klik opnieuw op de link in Ufora om opnieuw te proberen.
       success: "Succesvol geautoriseerd met %{kind} account."
+      redirect_to_known_provider: "Een gebruiker (%{user}) met dezelfde gebruikersnaam (%{username}) en/of hetzelfde email address (%{email}) bestaat al voor deze onderwijsinstelling (%{institution}). Als dit de gebruiker is waarin je probeerde in te loggen, bevestig dan je identiteit door in te loggen met een van onderstaande login methodes."
```
Comment (pdawyndt) [inline config/locales/devise.nl.yml:35]: de gebruiker waarVOOR je probeerde in te loggen

File: config/locales/devise.nl.yml [hunk 2]
```diff
@@ -32,6 +32,11 @@ nl:
...
+      redirect_to_known_provider_option: "Log in via %{provider}"
+      redirect_to_known_provider_contact: "Contacteer support als je problemen blijft hebben."
+      identity_created: "Vanaf nu kan je ook inloggen in deze gebruiker via %{provider}."
```
Comment (pdawyndt) [inline config/locales/devise.nl.yml:38]: VOOR deze gebruiker

File: config/locales/devise.nl.yml [hunk 3]
```diff
@@ -32,6 +32,11 @@ nl:
       wrong_provider_extra: "Aanmelden via %{user_provider_type}."
       linked: Omdat je nog niet eerder op Dodona inlogde vanaf Ufora, konden we de gevraagde actie niet uitvoeren. Nu we je UGent-account gekoppeld hebben, zou dit wel moeten lukken. Klik opnieuw op de link in Ufora om opnieuw te proberen.
       success: "Succesvol geautoriseerd met %{kind} account."
+      redirect_to_known_provider: "Een gebruiker (%{user}) met dezelfde gebruikersnaam (%{username}) en/of hetzelfde email address (%{email}) bestaat al voor deze onderwijsinstelling (%{institution}). Als dit de gebruiker is waarin je probeerde in te loggen, bevestig dan je identiteit door in te loggen met een van onderstaande login methodes."
+      redirect_to_known_provider_option: "Log in via %{provider}"
+      redirect_to_known_provider_contact: "Contacteer support als je problemen blijft hebben."
```
Comment (pdawyndt) [inline config/locales/devise.nl.yml:37]: Link naar contactformulier toevoegen?

File: app/views/users/show.html.erb [hunk 4]
```diff
@@ -86,6 +51,18 @@
...
           </div>
         </div>
+        <div class="row">
+          <h4><%= t(".sign_in") %></h4>
+          <% if @user.token.present? %>
+            <p><%= t ".token_sign_in" %></p>
+            <p><%= token_sign_in_user_url(@user, @user.token) %></p>
+          <% else %>
+            <span class="mb-2"><%= t ".providers_message"%></span>
+            <% @user.providers.where(mode: %i[prefer secondary]).each do |provider| %>
```
Comment (niknetniko) [inline app/views/users/show.html.erb:61]: On Naos, these buttons seem to link to `https://naos.ugent.be/nl/stats/punchcard?user_id=824` for me.

File: config/locales/devise.nl.yml [hunk 5]
```diff
@@ -32,6 +32,8 @@ nl:
       wrong_provider_extra: "Aanmelden via %{user_provider_type}."
       linked: Omdat je nog niet eerder op Dodona inlogde vanaf Ufora, konden we de gevraagde actie niet uitvoeren. Nu we je UGent-account gekoppeld hebben, zou dit wel moeten lukken. Klik opnieuw op de link in Ufora om opnieuw te proberen.
       success: "Succesvol geautoriseerd met %{kind} account."
+      identity_created: "Vanaf nu kan je ook inloggen voor deze gebruiker via %{provider}."
+      duplicate_email_for_provider: "Een andere gebruiker met hetzelfde email adres (%{email_address}) bestaat al in %{provider}. Er is een manuelle tussenkomst van het Dodona support team nodig om dit op te lossen."
```
Comment (niknetniko) [inline config/locales/devise.nl.yml:36]:
```suggestion
      duplicate_email_for_provider: "Een andere gebruiker met hetzelfde e-mailadres (%{email_address}) bestaat al in %{provider}. Er is een manuele tussenkomst van het Dodona support team nodig om dit op te lossen."
```

File: config/locales/views/auth/nl.yml [hunk 6]
```diff
@@ -15,3 +15,11 @@ nl:
       errors:
         institution-creation: Er ging iets mis tijdens het aanmaken van een nieuwe onderwijsinstelling.
         missing-provider: Enkel accounts die gelinkt zijn aan een onderwijsinstelling worden ondersteund.
+    redirect_to_known_provider:
+      title: Email adres en/of gebruikersnaam zijn niet uniek
```
Comment (niknetniko) [inline config/locales/views/auth/nl.yml:19]:
```suggestion
      title: E-mailadres en/of gebruikersnaam zijn niet uniek
```

File: config/locales/views/auth/en.yml [hunk 7]
```diff
@@ -15,3 +15,11 @@ en:
       errors:
         institution-creation: Something went wrong while creating a new institution.
         missing-provider: Only accounts linked to an institution are supported.
+    redirect_to_known_provider:
+      title: Email and/or username are not unique
+      message: "A user with the same username and/or email address already exists for this institution."
```
Comment (niknetniko) [inline config/locales/views/auth/en.yml:20]:
```suggestion
      message: "A user with the same email address and/or username already exists for this institution."
```

File: config/locales/views/auth/nl.yml [hunk 8]
```diff
@@ -15,3 +15,11 @@ nl:
...
+      message: "Een gebruiker met dezelfde gebruikersnaam en/of hetzelfde email adres bestaat al voor deze onderwijsinstelling."
```
Comment (niknetniko) [inline config/locales/views/auth/nl.yml:20]:
```suggestion
      message: "Een gebruiker met hetzelfde e-mailadres en/of dezelfde gebruikersnaam bestaat al voor deze onderwijsinstelling."
```

File: config/locales/views/auth/nl.yml [hunk 9]
```diff
@@ -15,3 +15,11 @@ nl:
...
+      providers_title: "Bevestig je identiteit"
+      contact_support_title: "Als jij dit niet bent, of je blijft problemen hebben"
+      contact_support: "Contacteer support"
+      user_title: "Gebruiker"
+      providers_message: "Als dit de gebruiker is waarvoor je probeerde in te loggen, bevestig dan je identiteit door in te loggen met een van onderstaande login methodes."
```
Comment (niknetniko) [inline config/locales/views/auth/nl.yml:25]:
```suggestion
      providers_message: "Als dit de gebruiker is waarvoor je probeerde in te loggen, bevestig dan je identiteit door in te loggen met een van onderstaande login-methodes."
```

File: config/locales/views/users/nl.yml [hunk 10]
```diff
@@ -39,6 +39,7 @@ nl:
       create_new_token: Nieuw token aanmaken
       token_help_html: API tokens laten andere programma's zoals de <a href="https://github.com/thepieterdc/dodona-plugin-jetbrains">Dodona plugin voor Jetbrains IDE's</a> toe om voor jou Dodona te gebruiken zonder je wachtwoord te moeten delen. Deze programma's zullen volledige toegang hebben tot jouw account, maar je kan die toegang gemakkelijk intrekken door het token te verwijderen onderaan deze pagina.
       description_help: De beschrijving is een geheugensteuntje zodat jij weet welk programma een bepaalde token gebruikt. Zo kan je eenvoudig de toegang van één programma kan ontzeggen zonder alle tokens te moeten verwijderen.
+      providers_message: "Je kan inloggen via onderstaande login methodes"
```
Comment (niknetniko) [inline config/locales/views/users/nl.yml:42]:
```suggestion
      providers_message: "Je kan inloggen via onderstaande login-methodes"
```

**Review by niknetniko — CHANGES_REQUESTED** [review]
This is more of a question, but is there an easy way to test this? Or would I have to modify the database on naos to adjust trigger this flow?

**Comment by jorg-vr** [comment]
@niknetniko I removed one of my identities on noas manually in the database to test this

File: app/views/users/show.html.erb (hunk 4 above)
Comment (jorg-vr) [reply app/views/users/show.html.erb:61]: Hmm odd issue, they do redirect to the sign in method. But if the user is you and you are already signed in it auto succeeds. 

After this it redirects to your last requested page. Which sounds logical, but in this case it is the punchcard. I'll have a look at fixing this

**Review by bmesuere — CHANGES_REQUESTED** [review]
Looks good 👍 

Can you remove the drop shadow from the providers on the profile page? They look a lot like buttons now which they aren't. I think outlined-cards could work here, but I don't think I implemented them yet in the m3 refactor.

**Review by bmesuere — APPROVED** [review]

File: config/locales/views/auth/nl.yml [hunk 11]
```diff
@@ -15,3 +15,10 @@ nl:
       errors:
         institution-creation: Er ging iets mis tijdens het aanmaken van een nieuwe onderwijsinstelling.
         missing-provider: Enkel accounts die gelinkt zijn aan een onderwijsinstelling worden ondersteund.
+    redirect_to_known_provider:
+      title: We hebben even je hulp nodig
+      message_p1: "Er bestaat al een gebruiker met dezelfde gebruikersnaam (**%{username}**) of hetzelfde e-mailadres (**%{email}**) binnen de onderwijsinstelling **%{institution}** op Dodona. Waarschijnlijk ben jij dit, maar gebruikte je vorige keer een andere loginmethode."
+      message_p2: "Om ook via **%{provider}** aan je gegevens te kunnen, moeten we even je bestaande account koppelen. Klik daarvoor op de aanmeldknop hieronder en log nogmaals in."
+      providers_title: "Accounts koppelen"
+      "contact_support_html": "Ben jij dit niet of blijf je problemen hebben? Vul het %{form} in en we helpen je zo snel mogelijk verder."
+      contact_form: "contact formulier"
```
Comment (chvp) [inline config/locales/views/auth/nl.yml:24]:
```suggestion
      contact_form: "contactformulier"
```

**Review by chvp — APPROVED** [review]

**Review by niknetniko — APPROVED** [review]
Looks good. I added a test I used to review this, and there was no reason to not commit it.

**Pull request merged**